core/src/main/java/org/apache/accumulo/core/client/impl/Table.java

```diff
@@ -16,24 +16,59 @@
  */
 package org.apache.accumulo.core.client.impl;
 
+import java.lang.ref.WeakReference;
+import java.util.WeakHashMap;
+
 import org.apache.accumulo.core.client.Instance;
 
 public class Table {
 
   /**
    * Object representing an internal table ID. This class was created to help with type safety. For help obtaining the value of a table ID from Zookeeper, see
    * {@link Tables#getTableId(Instance, String)}
+   *
+   * Uses an internal WeakHashMap and private constructor for storing a WeakReference of every Table.ID. Therefore, a Table.ID can't be instantiated outside
+   * this class and is accessed by calling Table.ID.{@link #of(String)}.
    */
   public static class ID extends AbstractId {
     private static final long serialVersionUID = 7399913185860577809L;
+    static final WeakHashMap<String,WeakReference<Table.ID>> tableIds = new WeakHashMap<>();
 
-    public static final ID METADATA = new ID("!0");
-    public static final ID REPLICATION = new ID("+rep");
-    public static final ID ROOT = new ID("+r");
+    public static final ID METADATA = of("!0");
+    public static final ID REPLICATION = of("+rep");
+    public static final ID ROOT = of("+r");
 
-    public ID(final String canonical) {
+    private ID(final String canonical) {
       super(canonical);
     }
+
+    /**
+     * Get a Table.ID object for the provided canonical string.
+     *
+     * @param canonical
+     *          table ID string
+     * @return Table.ID object
+     */
+    public static Table.ID of(final String canonical) {
+      return dedupeTableId(canonical);
+    }
+
+    private static Table.ID dedupeTableId(String tableIdString) {
```

Comment:
I thought about this but it would require reflection or passing a pre-constructed instance object as a parameter to the method.  I wanted to avoid reflection since we lose some of the type safety. I also wanted to avoid created objects every time for parameters since they may not be needed if the map already contains the id.  

Passing in a constructor object as a parameter may be a nice compromise but this introduces a bunch of potential exceptions:
<pre>
try {
        id = constructor.newInstance(idString);
      } catch (InstantiationException e) {
        e.printStackTrace();
      } catch (IllegalAccessException e) {
        e.printStackTrace();
      } catch (InvocationTargetException e) {
        e.printStackTrace();
      }
      idHashMap.put(idString, new WeakReference<>(id));
</pre>